src/backend/backend.c

```diff
@@ -301,108 +301,100 @@ void paint_all_new(session_t *ps, struct managed_win *t, bool ignore_damage) {
 			}
 
 			assert(w->shadow_image);
-			if (w->opacity == 1) {
-				ps->backend_data->ops->compose(
-				    ps->backend_data, w->shadow_image, w->g.x + w->shadow_dx,
-				    w->g.y + w->shadow_dy, &reg_shadow, &reg_visible);
-			} else {
-				auto new_img = ps->backend_data->ops->clone_image(
-				    ps->backend_data, w->shadow_image, &reg_visible);
-				ps->backend_data->ops->set_image_property(
-				    ps->backend_data, IMAGE_PROPERTY_OPACITY, new_img,
-				    &w->opacity);
-				ps->backend_data->ops->compose(
-				    ps->backend_data, new_img, w->g.x + w->shadow_dx,
-				    w->g.y + w->shadow_dy, &reg_shadow, &reg_visible);
-				ps->backend_data->ops->release_image(ps->backend_data, new_img);
-			}
+			ps->backend_data->ops->set_image_property(
+			    ps->backend_data, IMAGE_PROPERTY_OPACITY, w->shadow_image,
+			    &w->opacity);
+			ps->backend_data->ops->compose(
+			    ps->backend_data, w->shadow_image, w->g.x + w->shadow_dx,
+			    w->g.y + w->shadow_dy, &reg_shadow, &reg_visible);
 			pixman_region32_fini(&reg_shadow);
 		}
 
-		// Set max brightness
-		if (ps->o.max_brightness < 1.0) {
+		// Update image properties
+		{
+			double dim_opacity = ps->o.inactive_dim;
+			if (!ps->o.inactive_dim_fixed) {
+				dim_opacity *= w->opacity;
+			}
+
 			ps->backend_data->ops->set_image_property(
 			    ps->backend_data, IMAGE_PROPERTY_MAX_BRIGHTNESS, w->win_image,
 			    &ps->o.max_brightness);
+			ps->backend_data->ops->set_image_property(
+			    ps->backend_data, IMAGE_PROPERTY_INVERTED, w->win_image,
+			    &w->invert_color);
+			ps->backend_data->ops->set_image_property(
+			    ps->backend_data, IMAGE_PROPERTY_DIM_LEVEL, w->win_image,
+			    &dim_opacity);
```

Comment:
We have to check whether `w->dim` is actually set before setting the DIM_LEVEL. Otherwise all windows will be dimmed, regardless of their focus state.

Reply:
Thanks!